### PR TITLE
REFACTOR-#5363: introduce partition constructor; move `add_to_apply_calls` impl in base class

### LINKED ISSUE #5363: REFACTOR: introduce partition `__constructor__`
This will allow us to rewrite some code in partition hierarchy and avoid code duplication.

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5363 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/dataframe/pandas/partitioning/partition.py [hunk 1]
```diff
@@ -34,6 +34,18 @@ class PandasDataframePartition(ABC):  # pragma: no cover
     _width_cache = None
     _data = None
 
+    @property
```
Comment (vnlitvinov) [inline modin/core/dataframe/pandas/partitioning/partition.py:37]: we should probably decorate that with a `pandas.utils.cache_property` instead

**Review by vnlitvinov — APPROVED** [review]
LGTM!

**Comment by mvashishtha** [comment]
FYI This PR is also fixing #4821

**Review by vnlitvinov — APPROVED** [review]
LGTM!

**Comment by anmyachev** [comment]
@vnlitvinov @mvashishtha let's merge it

**Pull request merged**